File: 06/task2.c
// why can't lanternfish have natural predators ???? !!!
#include <stdlib.h>
#include <stdio.h>

#define MAX_AGE 10

int main()
{
    char file_name[] = "input.txt";
    // fill fish array with zeros
    unsigned long fish[MAX_AGE];
    for (int i = 0; i < MAX_AGE; i++)
        fish[i] = 0;
    
    FILE *fileptr = fopen(file_name, "r");
    // populate fish with input data
    char current_char = fgetc(fileptr);
    while (current_char != EOF && current_char != '\n') {
        if (current_char != ',')
            fish[current_char - '0']++;
        current_char = fgetc(fileptr);
    }

    int n = 255;
    unsigned long tmp[MAX_AGE];
    for (int i = 0; i < n; i++) {
        // reset tmp array
        for (int j = 0; j < MAX_AGE; j++)
            tmp[j] = 0;
        
        for (int k = 0; k < MAX_AGE; k++) {
            if (k != 0)
                tmp[k-1] = fish[k];
        }

        tmp[9] = tmp[0];
        tmp[7] += tmp[0];
        // copy tmp over to fish
        for (int i = 0; i < MAX_AGE; i++)
            fish[i] = tmp[i];
    }

    unsigned long sum = 0;
    for (int i = 0; i < MAX_AGE - 1; i++)
        sum += fish[i];

    printf("%ld\n", sum);

}

File: 06/task2.py
fish = []
with open("input.txt", "r") as f:
    fish = [int(item) if "\n" not in item else int(item[0:-1]) for item in f.readlines()[0].split(",")]


map = {0: 0, 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 7: 0, 8: 0, 9: 0}

for i in fish:
    map[i] += 1

n = 256 -1
for _ in range(n):
    #print(_, sum(map.values()) - map[9])
    tmp = {}
    for key, value in map.items():
        if key != 0:
            tmp[key-1] = value

    tmp[9] = tmp[0]
    tmp[7] += tmp[0]
    map = tmp

sum = 0
for key, val in map.items():
    if key != 9:
        sum += val
print(sum)
